flink-ml-iteration/src/test/java/org/apache/flink/iteration/operator/ReplayOperatorTest.java

```diff
@@ -50,12 +69,206 @@ public void testReplaying() throws Exception {
+        List<Object> secondRoundOutput = new ArrayList<>();
+
+        TaskStateSnapshot snapshot =
+                createHarnessAndRun(
+                        operatorId,
+                        null,
+                        harness -> {
+                            harness.getTaskStateManager().getWaitForReportLatch().reset();
+
+                            for (int i = 0; i < numRecords / 2; ++i) {
+                                harness.processElement(
+                                        new StreamRecord<>(IterationRecord.newRecord(i, 0)), 0, 0);
+                            }
+
+                            harness.getStreamTask()
+                                    .triggerCheckpointAsync(
+                                            new CheckpointMetaData(2, 1000),
+                                            CheckpointOptions.alignedNoTimeout(
+                                                    CheckpointType.CHECKPOINT,
+                                                    CheckpointStorageLocationReference
+                                                            .getDefault()));
+                            harness.processAll();
+
+                            firstRoundOutput.addAll(harness.getOutput());
+
+                            harness.getTaskStateManager().getWaitForReportLatch().await();
+                            return harness.getTaskStateManager()
+                                    .getLastJobManagerTaskStateSnapshot();
+                        });
+
+        createHarnessAndRun(
+                operatorId,
+                snapshot,
+                harness -> {
+                    for (int i = numRecords / 2; i < numRecords; ++i) {
+                        harness.processElement(
+                                new StreamRecord<>(IterationRecord.newRecord(i, 0)), 0, 0);
+                    }
+                    harness.endInput(0, true);
+                    harness.processElement(
+                            new StreamRecord<>(IterationRecord.newEpochWatermark(0, "send-0")),
+                            1,
+                            0);
+                    harness.processAll();
+                    firstRoundOutput.addAll(harness.getOutput());
+
+                    // The second round
+                    harness.getOutput().clear();
+                    harness.processElement(
+                            new StreamRecord<>(IterationRecord.newEpochWatermark(1, "send-0")),
+                            1,
+                            0);
+                    secondRoundOutput.addAll(harness.getOutput());
+
+                    return null;
+                });
+
+        assertOutputAllRecordsAndEpochWatermark(firstRoundOutput, numRecords, operatorId, 0);
+        assertOutputAllRecordsAndEpochWatermark(secondRoundOutput, numRecords, operatorId, 1);
+    }
+
+    @Test
+    public void testSnapshotAndRestoreOnSecondEpoch() throws Exception {
+        final int numRecords = 20;
```

Comment:
Would you like enlight me a little why this test case sets the `numRecords` 20 but others set it 10.

Reply:
It could also be 10, I changed it to 10 to be consistent.